### ISSUE TITLE
Reduce test suite to a more limited set of protocol version combinations

### ISSUE THREAD

**Issue by MonsieurNicolas** [stellar/stellar-core#1412]
Right now the approach for running tests is that we test all possible ranges of behavior for all versions of the protocol, which leads to an execution time proportional to the number of protocol versions.

While this guarantees that we're not breaking old versions of the protocol, the ROI seems very limited when put in context:
* ultimately, the bar applied for older versions of the protocol is: "can I catch-up to the current version of the protocol on the live network?" (and to answer this questions, we have the existing data set for the public network)
* with this in mind, we can reduce the range required to the version currently used on the network up to the most recent version supported by the binary (usually the difference between those 2 should be 1 as validators tend to be upgraded relatively quickly)
* for other versions, we want unit tests to have a high probability of catching problems breaking old versions.
   * this can be achieved by either picking a random version and/or when we know when a change occurred as part of a specific protocol version change, test the before/after for that protocol version

**Comment by vogel** [comment]
I'm using tests not only for testing, but also as a documentation of program behavior, this is why I added tests for each version of protocol.

For example: `for_all_versions` means that tested behavior didn't change at all, and `for_versions(1, 7)` means, that behavior was the same for versions 1,2,3,4,5,6,7.

I think that we could change range-based `for_versions` functions to only run on first and last element of range (or maybe on random element from that range)?

**Comment by MonsieurNicolas** [comment]
Yes, I think that in general using a random element instead of a full range is probably good enough in most cases. I think we'll need to use test coverage metrics to get a sense of how effective this approach would be.

As we'd be using more randomized tests, in order to avoid the "works on my machine" (or breaks in CI only) problem, we should probably take care of #1099 as part of this

**Comment by MonsieurNicolas** [comment]
@vogel this might be a way to work around the ever-increasing list of protocol versions. I forgot about this issue

**Comment by MonsieurNicolas** [comment]
@sisuresh @jonjove this is the issue I opened some time ago to discuss reducing CI time (we can do this as a follow up to the changes we're making for protocol 19)
